akka-actor-typed/src/main/scala/akka/actor/typed/internal/pubsub/TopicImpl.scala

```diff
@@ -0,0 +1,158 @@
+import akka.actor.typed.receptionist.ServiceKey
+import akka.actor.typed.scaladsl.AbstractBehavior
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.LoggerOps
+import akka.actor.typed.scaladsl.adapter._
+import akka.annotation.InternalApi
+
+import scala.reflect.ClassTag
+import scala.util.Random
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object TopicImpl {
+
+  trait Command[T]
+
+  final case class TopicInstancesUpdated[T](topics: Set[ActorRef[TopicImpl.Command[T]]]) extends Command[T]
+  final case class MessagePublished[T](message: T) extends Command[T]
+  final case class MessageSent[T](message: T) extends Command[T]
+  final case class SubscriberDied[T](subscriber: ActorRef[T]) extends Command[T]
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class TopicImpl[T](topicName: String, context: ActorContext[TopicImpl.Command[T]])(
+    implicit classTag: ClassTag[T])
+    extends AbstractBehavior[TopicImpl.Command[T]](context) {
+
+  /*
+   * The topic actor keeps a local set of subscribers, whenever that is non-empty it registers itself for
+   * a topic service key, and when it becomes empty again it deregisters from the service key. Published
+   * messages go to all currently known topics registered for the topic service key and the individual topic
+   * instances then broadcast the message to all local subscribers. This achieves deduplication for nodes
+   * with multiple subscribers and avoids sending to nodes without any subscribers.
+   */
+
+  import TopicImpl._
+
+  private val topicServiceKey = ServiceKey[TopicImpl.Command[T]](topicName)
+  context.log.debugN("Starting up pub-sub topic [{}] for messages of type [{}]", topicName, classTag.runtimeClass)
+
+  private var topics = Set.empty[ActorRef[TopicImpl.Command[T]]]
+  private var localSubscribers = Set.empty[ActorRef[T]]
+
+  private def receptionist = context.system.receptionist
+  private val receptionistAdapter = context.messageAdapter[Receptionist.Listing] {
+    case topicServiceKey.Listing(topics) => TopicInstancesUpdated(topics)
+  }
+  receptionist ! Receptionist.Subscribe(topicServiceKey, receptionistAdapter)
+
+  override def onMessage(msg: TopicImpl.Command[T]): Behavior[TopicImpl.Command[T]] = msg match {
+
+    case Topic.Publish(message) =>
+      if (topics.isEmpty) {
+        context.log.trace("Publishing message of type [{}] but no subscribers, dropping", msg.getClass)
+        context.system.deadLetters ! Dropped(message, "No topic subscribers known", context.self.toClassic)
+      } else {
+        context.log.trace("Publishing message of type [{}]", msg.getClass)
+        val pub = MessagePublished(message)
+        topics.foreach(_ ! pub)
+      }
+      this
+
+    case MessagePublished(msg) =>
+      context.log.trace("Message of type [{}] published", msg.getClass)
+      localSubscribers.foreach(_ ! msg)
+      this
+
+    case Topic.Send(message, preferLocal) =>
```

Comment:
does this have any advantages over using group router?

Reply:
It still has the thing where any number of subscribers per node only registers once to the receptionist, vs once per service. Not sure if that is motivation enough to keep though.

Reply:
That is for optimizing so that the message is only sent over the wire once, but for `Send` it will anyway only be one message (destination picked by the router on the sending side).

Reply:
Yep, I agree, removed `Send`